File: 20250421/leetcode2145. 统计隐藏数组数目/main.js
/*
 * @Author: mangwu                                                             *
 * @File: main.js                                                              *
 * @Date: 2025-04-21 23:26:19                                                  *
 * @LastModifiedDate: 2025-04-21 23:32:11                                      *
 * @ModifiedBy: mangwu                                                         *
 * -----------------------                                                     *
 * Copyright (c) 2025 mangwu                                                   *
 * -----------------------                                                     *
 * @HISTORY:                                                                   *
 * Date   	            By 	    Comments                                       *
 * ---------------------	--------	----------------------------------------------- *
 */

/**
 * @param {number[]} differences
 * @param {number} lower
 * @param {number} upper
 * @return {number}
 */
var numberOfArrays = function (differences, lower, upper) {
  // 求differences的前缀和，假设第一个元素是0
  let minNum = 0;
  let maxNum = 0;
  let cur = 0;
  for (const diff of differences) {
    cur += diff;
    minNum = Math.min(minNum, cur);
    maxNum = Math.max(maxNum, cur);
  }
  if (maxNum - minNum > upper - lower) return 0;
  return upper - lower - (maxNum - minNum) + 1;
};

File: 20250421/leetcode2145. 统计隐藏数组数目/main.cpp
#include <cmath>
#include <vector>

using namespace std;

class Solution {
public:
  int numberOfArrays(vector<int> &differences, int lower, int upper) {
    long long maxNum = 0;
    long long minNum = 0;
    long long cur = 0;
    for (int num : differences) {
      cur += num;
      maxNum = max(maxNum, cur);
      minNum = min(minNum, cur);
    }
    if (upper - lower < maxNum - minNum)
      return 0;
    return upper - lower - maxNum + minNum + 1;
  }
};